### PR TITLE
fix(plugin-compat): update fsevents patch to support virtual path

### PR DESCRIPTION
**Pull request by missing1984**
**What's the problem this PR addresses?**

Patch fsevents 2.x so it works with pnp virtual folders.

Close https://github.com/yarnpkg/berry/issues/1209

...

**How did you fix it?**

Find virtual paths for a given watch root and setup sub watchers accordingly.

...

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I have verified that all automated PR checks pass.

### CODE REVIEW

File: packages/plugin-compat/extra/fsevents/common.patch [hunk 1]
```diff
@@ -2,45 +2,108 @@ diff --git a/vfs.js b/vfs.js
...
+@@ -0,0 +1,104 @@
 +const path = require(`path`);
 +
-+const NUMBER_REGEXP = /^[0-9]+$/;
-+const VIRTUAL_REGEXP = /^(\/(?:[^\/]+\/)*?\$\$virtual)((?:\/([^\/]+)(?:\/([^\/]+))?)?((?:\/.*)?))$/;
++function getPnpApi() {
++  let pnpApi;
++  try {
++    if (process.versions.pnp) {
++      pnpApi = require(`pnpapi`);
```
Comment (missing1984) [inline packages/plugin-compat/extra/fsevents/common.patch:12]: @arcanis I am assuming calling `require(`pnpapi`)` many times won't have performance implication

File: packages/plugin-compat/extra/fsevents/1.2.11.patch [hunk 2]
```diff
@@ -2,20 +2,21 @@ diff --git a/fsevents.js b/fsevents.js
 semver exclusivity ^1
```
Comment (missing1984) [inline packages/plugin-compat/extra/fsevents/1.2.11.patch:2]: the API fsevents 1.x exposed makes it quite challenging to watch multiple files so I only watch the path file instead.( same behavior as current patch)

**Comment by arcanis** [comment]
I really want to look but the diff is difficult to read due to the GH interface not working super well for diffs 😞 I'll try to improve a bit our patch infra tomorrow to store the changes under JS syntax, and reapply your diff on top of it.

File: packages/plugin-compat/extra/fsevents/1.2.11.patch [hunk 3]
```diff
@@ -2,20 +2,21 @@ diff --git a/fsevents.js b/fsevents.js
...
++  var pathMap = vfs.buildPathMap();
++  var fse = new FSEvents(vfs.resolvedRoot, handler);
    EventEmitter.call(fse);
    return fse;
-
+ 
    function handler(path, flags, id) {
-+    path = vfs.transpose(path);
++    var rawRoot = pathMap.get(resolvedRoot);
++    path = vfs.transpose(rawRoot, resolvedRoot, path)
```
Comment (arcanis) [inline packages/plugin-compat/extra/fsevents/1.2.11.patch:19]: `resolvedRoot` is undefined here, I suspect you meant `vfs.resolvedRoot`

File: packages/plugin-compat/extra/fsevents/common.patch [hunk 4]
```diff
@@ -2,45 +2,108 @@ diff --git a/vfs.js b/vfs.js
...
 +  }
 +
-+  transpose(p) {
-+    return this.normalizedPath + p.substr(this.resolvedPath.length);
++  watch(handler) {
++    const pathMap = this.buildPathMap();
++    pathMap.forEach((virtualPath, resolvedPath) => {
++      const watcher = this.native.start(resolvedPath, (p, ...args) => {
++        return handler(this.transpose(virtualPath, resolvedPath, p), ...args);
++      });
```
Comment (arcanis) [inline packages/plugin-compat/extra/fsevents/common.patch:88]: I don't understand why it's necessary to spawn multiple watchers - the case you're trying to address is when calling `watch` on, for example, the root directory of the repository, right? If so, how would multiple watchers help?

**Comment by arcanis** [comment]
I think calling the watcher multiple times is problematic, as it will open multiple watchers on the same location. Since we already have reports that webpack sometimes open too many file descriptors, better be conservative here! I also would prefer to memoize as much work as possible, and to simultaneously avoid the `extname` call (which wouldn't work if the folder is called something like `node.js` or similar).

I've made a refactoring to instead compute an alias map . The advantage is that this map is shared between all calls, and it doesn't require much changes inside the fsevents files themselves. There's still room for perf improvements (I'm currently doing a crawl on all the options, which isn't ideal - a binary tree would be more efficient), but that should work without too much code.

Can you check whether it works for you? I've added a test for both FSEvents 1 & 2, but better be sure 😄

**Pull request merged**